bundles/org.openhab.binding.weathercompany/src/main/java/org/openhab/binding/weathercompany/internal/handler/WeatherCompanyHandler.java

```diff
@@ -0,0 +1,744 @@
+    private @NonNullByDefault({}) HttpClient httpClient;
+    private @NonNullByDefault({}) SystemOfUnits systemOfUnits;
+
+    // Thing configuration
+    private @NonNullByDefault({}) WeatherCompanyConfig config;
+
+    private @Nullable String forecastUrl;
+    private @Nullable String pwsUrl;
+
+    // Job to update the forecast and PWS observations
+    private @Nullable Future<?> refreshJob;
+    private int refreshInterval;
+
+    private Runnable refreshRunnable = new Runnable() {
+        @Override
+        public void run() {
+            refreshForecast();
+            refreshPwsObservations();
+        }
+    };
+
+    public WeatherCompanyHandler(Thing thing, TimeZoneProvider timeZoneProvider, HttpClient httpClient,
+            UnitProvider unitProvider) {
+        super(thing);
+        this.timeZoneProvider = timeZoneProvider;
+        this.httpClient = httpClient;
+        this.systemOfUnits = unitProvider.getMeasurementSystem();
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+
+        // Get the configuration
+        config = getConfigAs(WeatherCompanyConfig.class);
+        // logger.debug("Handler: Config: apiKey={}", config.apiKey);
+        logger.debug("Handler: Config: locationType={}", config.locationType);
+        logger.debug("Handler: Config: postalCode={}", config.postalCode);
+        logger.debug("Handler: Config: geocode={}", config.geocode);
+        logger.debug("Handler: Config: language={}", config.language);
+        logger.debug("Handler: Config: pwsStationId={}", config.pwsStationId);
+        logger.debug("Handler: Config: refreshInterval={} minutes", config.refreshInterval);
+
+        // Set the frequency with which the forecast will be refreshed
+        refreshInterval = getRefreshIntervalSeconds();
+
+        // Construct the URL for querying the forecast
+        forecastUrl = buildForecastUrl();
+
+        // Construct the URL for querying the PWS observations
+        pwsUrl = buildPwsUrl();
+
+        // Schedule the job to refresh the forecast
+        scheduleRefreshJob();
+    }
+
+    @Override
+    public void dispose() {
+        cancelRefreshJob();
+        updateStatus(ThingStatus.OFFLINE);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // Handler doesn't handle any commands
+    }
+
+    /*
+     * Build the URL for requesting the 5-day forecast
+     */
+    private String buildForecastUrl() {
+        StringBuilder sb = new StringBuilder(BASE_FORECAST_URL);
+        // Set response type as JSON
+        sb.append("?format=json");
+        // Set language from config
+        sb.append("&language=").append(config.language);
+        // Set API key from config
+        sb.append("&apiKey=").append(config.apiKey);
+        // Set the units to Imperial or Metric
+        sb.append("&units=").append(getUnitsQueryString());
```

Comment:
Why do you pass the unit here but not on line 245 is that something which should be done, or are they both correct?

Reply:
Good spot! This was intentional.

For the PWS observations, it's sufficient to request Imperial units, then let UoM convert to the units specified in the user's system config.

However, there's a bit of an oddity for the forecast. 

Normally, I would handled the forecast the same way I handled the PWS observations, i.e. request Imperial, then let UoM handle the units conversion based on the user's config. But, since the narrative contains values (e.g. wind speed) as well as text (e.g. "wind gusting to 30 mph"), a mismatch of language and units can cause the narrative to be inconsistent.

In order for the narrative to contain the correct language AND value, you must call the API with the intended language AND units. Otherwise you get something like "Windböen bis 30 mph" or "Wind gusts up to 48 kph".

I've updated the comment.